torch/csrc/autograd/variable.h

```diff
@@ -106,8 +106,11 @@ struct TORCH_API Variable : public at::Tensor {
 
   /// Creates a `Variable` from the given `Tensor`. `requires_grad` should be
   /// set only for leaves, and determines whether the `Variable` will accumulate
-  /// gradients. NOTE: `data` must *not* be a `Variable` already. Its dynamic
-  /// type *must* be `Tensor`.
+  /// gradients.
+  ///
+  /// NOTE: if the given `Tensor` is already a `Variable`, this function will
+  /// create a new `Variable` that shares the same storage and tensor metadata
+  /// with the original `Variable`, but with a completely new autograd history.
```

Comment:
Thought: is it perhaps better to error if we pass a Variable with history? What happens if I do `x.data = torch.zeros(2, requires_grad=True)` in Python today?

Reply:
If we do `x.data = torch.zeros(2, requires_grad=True)` in Python today, `x.requires_grad` would be false, because `var.set_data(tensor)` does not copy AutogradMeta from `tensor` to `var`.

I think we should accept Variable with history for this factory function, because it will be useful for call sites that want to create a new Variable that shares the same storage and tensor metadata with another Variable but with new autograd history (e.g. `THPVariable_get_data`). Also the semantics for the `make_variable()` factory functions have always been that the original autograd history is ignored and a new history is always created, so we are not breaking the original semantics here.

Reply:
`make_variable()` is useful for call sites that needs to shallow-copy a variable and create a new autograd history, e.g in `THPVariable_get_data`: 
```cpp
static PyObject * THPVariable_get_data(THPVariable *self)
{
  HANDLE_TH_ERRORS
  /// NOTE: Previously, if we change the tensor metadata (e.g. sizes / strides /
  /// storage / storage_offset) of a tensor created from `.data`, those metadata
  /// in the original tensor will also be updated. However, the new behavior is that
  /// those metadata changes to the `.data` tensor will not update the original tensor
  /// anymore, and here we need to set `allow_tensor_metadata_change_` to false to
  /// make such changes explicitly illegal, in order to prevent users from changing
  /// metadata of the `.data` tensor and expecting the original tensor to also
  /// be updated.
  auto var = make_variable(self->cdata, /*requires_grad=*/false, /*allow_tensor_metadata_change=*/false);
  return THPVariable_Wrap(var);
  END_HANDLE_TH_ERRORS
}
```
I will rename `make_variable()` to something similar to `shallow_copy_with_new_autograd_metadata()` in my next PR (setting AutogradMeta only when `requires_grad=true`).

Reply:
I still don't get it.

As far as I see it, there are two cases when this function is called:
1) when we have a (historical) Tensor, something without an autograd_meta, where we want to create an autograd_meta because we need it.
2) when we essentially want to call `.data` from Python, but that API does something completely different in C++, so we do something like `make_variable(tensor.data(), false, false)`.  But this is a completely different use case! 

Instead of trying to combine these use cases, just make different functions.  2) will eventually become `.data()` -- just give it a different name for now.